Web components 101 (#15)
nifty10m/devsquad: _posts/2019-06-11-web-components-101.md

In order to get the `attributeChangedCallback` to fire you have to define the attributes you want to observe by adding a static getter to the class.

```
class MyButton extends HTMLButtonElement {
  static get observedAttributes() {
    return ['width', 'height'];
  }
}
```

Lastly you have to register your web component with customElements global object by calling `customElements.define()`. The define method expects a class extending HTMLElement and a tag name containing at least one dash for namespacing your element.
And that is your first web component done!

## Why should I create/use web components?

Web components provide all the benefits of any component based architectures:
* Composability
* Reusability
* Maintainability
* Extensibility
* Productivity

And some additional benefits because they are used in the web:
* Scoping
* Interoperability (framework agnostic)
* Accessibility (by extending default HTML elements)

Interoperability is especially important, since this allows you easy brand consistency. Create your corporate components library or design system as web components

> and all your teams can work off the same base no matter which framework they use. This will also directly reflect in higher development speed because

Reviewer: ```suggestion
and all teams can work on the same base no matter which framework they use. This will also directly reflect in higher development speed because
```

the components don't have to be implemented for each framework over and over again.
Lastly web components is a browser standard added by the W3C which means it uses core features of each browser.

However, there are certain downsides to web components that justify the coexistence of the known web frameworks.

## What can't I do with web components?

Web components have inputs and outputs (events) much like any other framework but primarily the inputs are pretty limited for they only
accept strings. Which means you can pass in types that can be cast to string or can be deserialized from JSON at most.
Also reacting to changes of input values is rather tedious to setup since you have to create a separate listener for each attribute you want to react on.

> All in all I would recommend using web components for simple leaf nodes (e.g. buttons, inputs, etc.) or certain layout components (e.g. a bootstrap grid system).

Reviewer: ```suggestion
All together I would recommend using web components for simple leaf nodes (e.g. buttons, inputs, etc.) or certain layout components (e.g. a bootstrap grid system).
```
